test/src/unit-vfs.cc

```diff
@@ -351,3 +352,86 @@ TEST_CASE("VFS: test ls_with_sizes", "[vfs][ls-with-sizes]") {
   // Clean up
   REQUIRE(vfs_ls.remove_dir(URI(path)).ok());
 }
+
+// Currently only S3 is supported for VFS::ls_cb.
+using TestBackends = std::tuple<S3Test>;
+TEMPLATE_LIST_TEST_CASE(
+    "VFS: Test internal ls_cb recursion argument",
+    "[vfs][ls_cb]",
+    TestBackends) {
+  TestType fs({10, 50});
+  if (!fs.is_supported()) {
+    return;
+  }
+
+  tiledb::sm::LsCallbackCAPI cb =
+      [](const char* path, size_t path_len, uint64_t size, void* data) {
+        auto ls_objects = static_cast<VFSTest::LsObjects*>(data);
+        ls_objects->emplace_back(std::string(path, path_len), size);
+        return 1;
+      };
+
+  bool recursive = GENERATE(false, true);
+  DYNAMIC_SECTION(
+      fs.temp_dir_.backend_name()
+      << " ls_cb with recursion: " << (recursive ? "true" : "false")) {
+    fs.test_ls_cb(cb, recursive);
```

Comment:
This statement uses a test that's defined completely elsewhere. That's generally a sign that the code is organized poorly. The code that runs this test should be in the present source file (unless it's shared by other units, which it's not in this case).